File: webapp/lib/bait_icons.rb
# frozen_string_literal: true

require_relative "bait"

# An immutable count of icons per bait type. Answers "how many of bait X?".
# A bait type that is absent counts as 0.
class BaitIcons
  # @param counts [Hash] raw map of bait name => count (from YAML or code)
  def initialize(counts = {})
    @counts = {}
    (counts || {}).each do |bait, count|
      next if count.to_i.zero?

      @counts[Bait.normalize(bait)] = Integer(count)
    end
    @counts.freeze
  end

  # Number of icons of the given bait type.
  def count(bait)
    @counts.fetch(Bait.normalize(bait), 0)
  end

  # Total icons across all bait types.
  def total
    @counts.values.sum
  end

  def to_h
    @counts.dup
  end

  # True if this set has at least as many icons of every type as `other`
  # (used to check an advanced room's bait requirement against a room).
  def contains?(other)
    other.to_h.all? { |bait, needed| count(bait) >= needed }
  end

  # True if this set and `other` have at least one bait type in common (used to
  # check an advanced room's bait requirement: it may replace any room that
  # shares at least one of its bait icons).
  def shares?(other)
    (to_h.keys & other.to_h.keys).any?
  end
end

File: android/app/src/main/java/com/dungeonboss/model/BaitIcons.kt
package com.dungeonboss.model

/**
 * An immutable count of icons per bait type. Answers "how many of bait X?".
 * A bait type that is absent counts as 0. Accepts either a raw map from YAML
 * (string keys) or a [Bait]-keyed map; zero counts are dropped.
 */
class BaitIcons(counts: Map<*, *>? = null) {
    private val counts: Map<Bait, Int>

    init {
        val result = LinkedHashMap<Bait, Int>()
        counts?.forEach { (bait, count) ->
            val n = when (count) {
                is Number -> count.toInt()
                null -> 0
                else -> count.toString().toIntOrNull() ?: 0
            }
            if (n != 0) result[Bait.normalize(bait)] = n
        }
        this.counts = result
    }

    /** Number of icons of the given bait type. */
    fun count(bait: Bait): Int = counts[bait] ?: 0

    /** Total icons across all bait types. */
    fun total(): Int = counts.values.sum()

    /** True when, for every bait in [other], this holds at least as many. */
    fun contains(other: BaitIcons): Boolean =
        other.counts.all { (bait, n) -> count(bait) >= n }

    /** True when any bait type appears (>0) in both. */
    fun shares(other: BaitIcons): Boolean =
        counts.keys.any { other.count(it) > 0 }

    fun toMap(): Map<Bait, Int> = LinkedHashMap(counts)
}
